src/validation.cpp

```diff
@@ -372,11 +375,16 @@ void Chainstate::MaybeUpdateMempoolForReorg(
             }
         }
         // Transaction is still valid and cached LockPoints are updated.
+        auto ancestors{m_mempool->AssumeCalculateMemPoolAncestors(func, *it, CTxMemPool::Limits::NoLimits(),
```

Comment:
1) This strikes me as a very expensive operation to perform in order to update the mempool for a reorg.  I think we could probably do better by only calculating ancestors or descendants for transactions that are themselves v3, if we decide we need to enforce the v3 rules after a reorg.

2) With the logic as written, if a v3 parent-child pair (A, B) were to be reorged out and added back to the mempool, and A has some other spend C already in the mempool, then when `ApplyV3Rules()` is invoked on B, it would be selected for removal.  This in turn would prevent A from being mined, if we assume that B's feerate was the reason it was mined in the first place?  This strikes me as unfortunate.

Instead, I'd suggest that we not bother enforcing the v3 rules on a reorg.  That seems safer than potentially evicting the fee-paying transaction, less CPU work (so reorgs won't be further slowed down by additional processing), and shouldn't materially affect pinning (if miners are reorging out your transactions, trying to minimize RBF pinning with v3 policy isn't really going to help you!).

Reply:
> Instead, I'd suggest that we not bother enforcing the v3 rules on a reorg.

I'm a fan of this approach. I want to spend a little bit of time thinking about what the possibilities are - I guess we can have some incorrect assumptions about transactions (resulting in e.g. incorect MinerScores) but it wouldn't be problematic if a reorg is the only way to get there, as you say.